R/parse-body.R

```diff
@@ -16,230 +12,461 @@ postBodyFilter <- function(req){
+  function(value, filename = "", ...) {
     tmp <- tempfile("plumb", fileext = paste0("_", basename(filename)))
-    on.exit(file.remove(tmp), add = TRUE)
+    on.exit({
+      file.remove(tmp)
+    }, add = TRUE)
     writeBin(value, tmp)
-    readRDS(tmp)
+    read_fn(tmp)
+  }
+}
+
+#' @describeIn parsers CSV parser
+#' @export
+parser_csv <- function(...) {
+  if (!requireNamespace("readr", quietly = TRUE)) {
+    stop("`readr` must be installed for `parser_csv` to work")
   }
+  parser_read_file(function(tmpfile) {
+    readr::read_csv(tmpfile, ...)
+  })
+}
+
+
+#' @describeIn parsers TSV parser
+#' @export
+parser_tsv <- function(...) {
+  if (!requireNamespace("readr", quietly = TRUE)) {
+    stop("`readr` must be installed for `parser_tsv` to work")
+  }
+  parser_read_file(function(tmpfile) {
+    readr::read_tsv(tmpfile, ...)
+  })
+}
+
+
+#' @describeIn parsers RDS parser
+#' @export
+parser_rds <- function(...) {
+  parser_read_file(function(tmpfile) {
+    readRDS(tmpfile, ...)
+  })
 }
 
 
+#' @describeIn parsers Octet stream parser. Will add a filename attribute if the filename exists
+#' @export
+parser_octet <- function() {
+  function(value, filename = NULL, ...) {
+    attr(value, "filename") <- filename
+    value
+  }
+}
 
 
-#" MULTI
-#' @rdname parsers
+#' @describeIn parsers Multi part parser. This parser will then parse each individual body with its respective parser
 #' @export
 #' @importFrom webutils parse_multipart
 parser_multi <- function() {
-  function(value, content_type, ...) {
+  function(value, content_type, parsers, ...) {
     if (!stri_detect_fixed(content_type, "boundary=", case_insensitive = TRUE))
       stop("No boundary found in multipart content-type header: ", content_type)
     boundary <- stri_match_first_regex(content_type, "boundary=([^; ]{2,})", case_insensitive = TRUE)[,2]
     toparse <- parse_multipart(value, boundary)
     # content-type detection
     lapply(toparse, function(x) {
-      if (!is.null(x$filename)) {
-        x$content_type <- getContentType(tools::file_ext(x$filename))
+      if (
+        is.null(x$content_type) ||
+        # allows for files to be shipped as octect, but parsed using the matching value in `knownContentTypes`
+        # (Ex: `.rds` files -> `application/rds` which has a proper RDS parser)
+        isTRUE(stri_detect_fixed(x$content_type, "application/octet-stream"))
+      ) {
+        if (!is.null(x$filename)) {
+          # Guess content-type from file extension
+          x$content_type <- getContentType(tools::file_ext(x$filename))
+        }
       }
-      parseRaw(x)
+      x$parsers <- parsers
+      parse_raw(x)
     })
   }
 }
 
-
-
-
-#' OCTET
-#' @rdname parsers
+#' @describeIn parsers Enable all parsers. Not recommended due to security concerns.
 #' @export
-parser_octet <- function() {
-  function(value, filename = NULL, ...) {
-    attr(value, "filename") <- filename
-    return(value)
+parser_all <- function() {
```

Comment:
I'm not sure you would need an all and a none parser since it could be all managed from make_parsers.
Also parser_all is broken

make_parsers("none") or make_parsers(character())
make_parsers("all")

Reply:
Correct. Moved the logic inside for the `"all"` case.

I still want to maintain the `parser_none` to be a function.  This maintains the contract that a `parser` returns a function that can parse a value.  The only one that breaks this contract is `"all"`, which is now handled by `make_parser("all")`.